vendor/bam/src/json/response.rs

```diff
@@ -1,12 +1,15 @@
 use crate::{
     api::{self, IntoFrame},
-    json::frame::Frame,
+    json::{self, frame::Frame},
 };
-use serde::Serialize;
-use serde_json::{self, Value as JsonValue};
-use std::collections::HashMap;
+use serde::{
+    de::{self, Deserialize, DeserializeOwned, Deserializer, MapAccess, Visitor},
+    Serialize,
+};
+use serde_json::{self, Map, Value as JsonValue};
+use std::{collections::HashMap, fmt};
 
-#[derive(Serialize, Deserialize, Debug, PartialEq)]
+#[derive(Serialize, Debug, PartialEq)]
```

Comment:
Huh? What was wrong with the derived Deserialization?
Please note that the way things worked previously is a total mess altogether and still the early PoC that Lloyd and I put together in like 2 weeks.

The whole BAM crate needs a some love through better modelling and use of the type system. Atm we have too many uses of serde_json::Value that we interpret at runtime.

Reply:
This stupid `deserialization_failed` flag for example is never checked :D 

You (@luckysori) seem to be interested in working on this. Have a go and make it more typesafe and less runtime-error prone :)

Reply:
The problem with the old derived deserialization is that it couldn't work for both compact headers such as
```
headers: {
  "HELLO": "WORLD"
}
```
and full headers like
```
headers: {
  "HELLO": {
    "value": "WORLD"
  }
}
```
Maybe there's an alternative way of doing this without having to define the deserialization in full, but I couldn't figure it out.